[java] ParseException occurs with valid '<>' in Java 1.8 mode

**Affects PMD Version:** 
6.0.0

**Rule:**

**Description:**
A valid '<>' (diamond operator) in java 1.8 is misrecognized as anonymous inner class  and a ParseException occurs.

**Code Sample demonstrating the issue:**
```
import java.awt.*;
import javax.swing.*;
import javax.swing.plaf.LayerUI;
import javax.swing.plaf.basic.BasicComboBoxEditor;

public class PMDExceptionTest {
  private Component makeUI() {
    String[] model = {"123456", "7890"};
    JComboBox<String> comboBox = new JComboBox<>(model);
    comboBox.setEditable(true);
    comboBox.setEditor(new BasicComboBoxEditor() {
      private Component editorComponent;
      @Override public Component getEditorComponent() {
        if (editorComponent == null) {
          JTextField tc = (JTextField) super.getEditorComponent();
          editorComponent = new JLayer<>(tc, new ValidationLayerUI<>());
        }
        return editorComponent;
      }
    });
    JPanel p = new JPanel();
    p.add(comboBox);
    return p;
  }
  public static void main(String[] args) {
    EventQueue.invokeLater(() -> {
      JFrame f = new JFrame();
      f.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
      f.getContentPane().add(new PMDExceptionTest().makeUI());
      f.setSize(320, 240);
      f.setLocationRelativeTo(null);
      f.setVisible(true);
    });
  }
}
class ValidationLayerUI<V extends JTextField> extends LayerUI<V> {
  @Override public void paint(Graphics g, JComponent c) {
    super.paint(g, c);
    // paint an icon and so on but omit it
  }
}
```

```
$ "%JAVA_HOME%/bin/javac" -version
javac 1.8.0_162

$ "%JAVA_HOME%/bin/javac" src/PMDExceptionTest.java

$ dir src
'PMDExceptionTest$1.class'   PMDExceptionTest.class   PMDExceptionTest.java   ValidationLayerUI.class

$ "%PMD_HOME%/bin/pmd.bat" -debug -d ./src -R category/java/bestpractices.xml -language java -version 1.8
Loaded rule ...
...
Processing C:\temp\src\PMDExceptionTest.java
Error while processing file: C:\temp\src\PMDExceptionTest.java
net.sourceforge.pmd.PMDException: Error while parsing C:\temp\src\PMDExceptionTest.java
        at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:98)
        at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:50)
        at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:75)
        at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:23)
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: net.sourceforge.pmd.lang.java.ast.ParseException: Line 21, Column 5: Cannot use '<>' with anonymous inner classes when running in JDK inferior to 9 mode!
        at net.sourceforge.pmd.lang.java.ast.JavaParser.throwParseException(JavaParser.java:23)
        at net.sourceforge.pmd.lang.java.ast.JavaParser.checkForBadAnonymousDiamondUsage(JavaParser.java:139)
        at net.sourceforge.pmd.lang.java.ast.JavaParser.AllocationExpression(JavaParser.java:4519)
        at net.sourceforge.pmd.lang.java.ast.JavaParser.PrimaryPrefix(JavaParser.java:3922)
```

**Running PMD through:** *[CLI]*

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -29,6 +29,7 @@ This is a minor release.
     *   [#791](https://github.com/pmd/pmd/issues/791): \[doc] Documentation site reorganisation
 *   java
     *   [#825](https://github.com/pmd/pmd/issues/825): \[java] Excessive\*Length ignores too much
+    *   [#888](https://github.com/pmd/pmd/issues/888): \[java] ParseException occurs with valid '<>' in Java 1.8 mode
 *   java-documentation
     *   [#877](https://github.com/pmd/pmd/issues/877): \[java] CommentRequired valid rule configuration causes PMD error
 *   java-errorprone

--- a/pmd-java/etc/grammar/Java.jjt
+++ b/pmd-java/etc/grammar/Java.jjt
@@ -1,4 +1,7 @@
 /**
+ * Fixes #888 [java] ParseException occurs with valid '<>' in Java 1.8 mode
+ * Juan Martin Sotuyo Dodero 01/2018
+ *====================================================================
  * Fixes #793 [java] Parser error with private method in nested classes in interfaces
  * Andreas Dangel 12/2017
  *====================================================================
@@ -333,9 +336,9 @@ public class JavaParser {
   private void checkForBadAnonymousDiamondUsage() {
     if (jdkVersion < 9) {
       ASTAllocationExpression node = (ASTAllocationExpression)jjtree.peekNode();
-      ASTTypeArguments types = node.getFirstDescendantOfType(ASTTypeArguments.class);
+      ASTTypeArguments types = node.getFirstChildOfType(ASTClassOrInterfaceType.class).getFirstChildOfType(ASTTypeArguments.class);
       if (node.isAnonymousClass() && types != null && types.isDiamond()) {
-          throwParseException("Cannot use '<>' with anonymous inner classes when running in JDK inferior to 9 mode!");
+        throwParseException("Cannot use '<>' with anonymous inner classes when running in JDK inferior to 9 mode!");
       }
     }
   }
@@ -2026,8 +2029,8 @@ void AllocationExpression():
             { inInterface = inInterfaceOld; } // always restore the flag after leaving the node
         ]
      )
+    { checkForBadAnonymousDiamondUsage(); }
   )
-  { checkForBadAnonymousDiamondUsage(); }
 }
 
 /*

--- a/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/ParserCornersTest.java
+++ b/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/ParserCornersTest.java
@@ -35,6 +35,33 @@ public class ParserCornersTest {
                 + "                TestInnerClassCallsOuterParent.super.toString();\n" + "            }\n"
                 + "        };\n" + "    }\n" + "}\n");
     }
+    
+    /**
+     * #888 PMD 6.0.0 can't parse valid <> under 1.8.
+     */
+    @Test
+    public void testDiamondUsageJava8() {
+        parseJava18("public class PMDExceptionTest {\n" + 
+                "  private Component makeUI() {\n" + 
+                "    String[] model = {\"123456\", \"7890\"};\n" + 
+                "    JComboBox<String> comboBox = new JComboBox<>(model);\n" + 
+                "    comboBox.setEditable(true);\n" + 
+                "    comboBox.setEditor(new BasicComboBoxEditor() {\n" + 
+                "      private Component editorComponent;\n" + 
+                "      @Override public Component getEditorComponent() {\n" + 
+                "        if (editorComponent == null) {\n" + 
+                "          JTextField tc = (JTextField) super.getEditorComponent();\n" + 
+                "          editorComponent = new JLayer<>(tc, new ValidationLayerUI<>());\n" + 
+                "        }\n" + 
+                "        return editorComponent;\n" + 
+                "      }\n" + 
+                "    });\n" + 
+                "    JPanel p = new JPanel();\n" + 
+                "    p.add(comboBox);\n" + 
+                "    return p;\n" + 
+                "  }\n" + 
+                "}");
+    }
 
     @Test
     public final void testGetFirstASTNameImageNull() {

--- a/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/ParserCornersTest.java
+++ b/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/ParserCornersTest.java
@@ -41,26 +41,26 @@ public class ParserCornersTest {
      */
     @Test
     public void testDiamondUsageJava8() {
-        parseJava18("public class PMDExceptionTest {\n" + 
-                "  private Component makeUI() {\n" + 
-                "    String[] model = {\"123456\", \"7890\"};\n" + 
-                "    JComboBox<String> comboBox = new JComboBox<>(model);\n" + 
-                "    comboBox.setEditable(true);\n" + 
-                "    comboBox.setEditor(new BasicComboBoxEditor() {\n" + 
-                "      private Component editorComponent;\n" + 
-                "      @Override public Component getEditorComponent() {\n" + 
-                "        if (editorComponent == null) {\n" + 
-                "          JTextField tc = (JTextField) super.getEditorComponent();\n" + 
-                "          editorComponent = new JLayer<>(tc, new ValidationLayerUI<>());\n" + 
-                "        }\n" + 
-                "        return editorComponent;\n" + 
-                "      }\n" + 
-                "    });\n" + 
-                "    JPanel p = new JPanel();\n" + 
-                "    p.add(comboBox);\n" + 
-                "    return p;\n" + 
-                "  }\n" + 
-                "}");
+        parseJava18("public class PMDExceptionTest {\n"
+                + "  private Component makeUI() {\n"
+                + "    String[] model = {\"123456\", \"7890\"};\n"
+                + "    JComboBox<String> comboBox = new JComboBox<>(model);\n"
+                + "    comboBox.setEditable(true);\n"
+                + "    comboBox.setEditor(new BasicComboBoxEditor() {\n"
+                + "      private Component editorComponent;\n"
+                + "      @Override public Component getEditorComponent() {\n"
+                + "        if (editorComponent == null) {\n"
+                + "          JTextField tc = (JTextField) super.getEditorComponent();\n"
+                + "          editorComponent = new JLayer<>(tc, new ValidationLayerUI<>());\n"
+                + "        }\n"
+                + "        return editorComponent;\n"
+                + "      }\n"
+                + "    });\n"
+                + "    JPanel p = new JPanel();\n"
+                + "    p.add(comboBox);\n"
+                + "    return p;\n"
+                + "  }\n"
+                + "}");
     }
 
     @Test

